### DISCUSSION TITLE
Link Margin Calculations in Noisy Environments

### DISCUSSION THREAD

**Discussion by mmb-davidsmith** [openthread/openthread#10571] [Q&A] [answered]
I've been working with a customer recently to try to understand and improve the performance of their devices in noisy environments. Once of the items that has been discussed is the link margin calculation as it has downstream impacts on link quality and the associated route metrics.

In the thread spec, it says that the link margin is calculated as the `Link Margin = RSSI - noise_floor`. It then says that the 

> noise_floor is [one] of the following:
>  * measured noise floor
>  OR 
>  * receiver sensitivity
>  OR
>  * constant

For the devices in question (all EFR32 family based, though different MCUs between router, sleepy sensors and data aggregator), the stack has chosen to go with a constant receiver sensitivity of -100. This means that if I see the RSSI of -75 from a device, that would calculate as a link margin of 25 which provides a link quality of 3.

In my development environment (which is similar to their deployment environment), if I run `scan energy 100 <channel>`, I tend to see a reported RSSI that varies between -80 and -70 as I read it over time. My understanding is that this essentially gives me a picture of what the noise on a given channel looks like when this is run.

 If I'm understanding the purpose and results of the energy scan correct, that means that I could sample that periodically over time and use a exponentially weighted moving average, a mean across N most recent samples, or a median across some set on N most recent samples to determine a `measured noise floor`. In the case of my dev environment, I manually took 10 samples at 1 second intervals and got the following results (first is the newest value):
```
[-78, -80, -79, -80, -80, -80, -80, -70, -79, -79]
```

This gives a mean of -78.5 (let's round generously to -79). Wit this, we get a link margin of 4, which would give a link quality of 1. This is a big difference, and would likely affect network reliability a fair bit in very noisy environments.

Could you confirm if I'm understanding this correctly? If it is, I'll likely experiment with trying to implement some sort of noise floor monitoring in my routers to see if it improves the overall network reliability.

**Comment by jwhui** [comment, marked as answer]
Yes, your understanding is correct.

The [Channel Monitor component](https://github.com/openthread/openthread/blob/main/src/core/utils/channel_monitor.cpp) does something similar using energy scan, but across all channels. Maybe you will find it useful, at least as an example of using energy scan.
